skydb/skydb.go

```diff
@@ -0,0 +1,159 @@
+package skydb
+
+import (
+	"bytes"
+	"crypto/sha1"
+	"encoding/base64"
+	"os"
+	"strings"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/SkynetLabs/skyd/node/api/client"
+	"gitlab.com/SkynetLabs/skyd/skymodules"
+	"go.sia.tech/siad/crypto"
+	"go.sia.tech/siad/modules"
+	"go.sia.tech/siad/types"
+)
+
+var ErrNotFound = errors.New("entry not found")
+
+// SkyDBI is the interface for communicating with SkyDB. We use an interface, so
+// we can easily override it for testing purposes.
+type SkyDBI interface {
+	Read(crypto.Hash) ([]byte, uint64, error)
+	Write(data []byte, dataKey crypto.Hash, rev uint64) error
+}
+
+type SkyDB struct {
+	Client *client.Client
+	sk     crypto.SecretKey
+	pk     crypto.PublicKey
+}
+
+// New creates a new SkyDB client with default option.
+// The default options are to look for a skyd node at localhost:9980,
+// to use "Sia-Agent" as user agent and to get the password for skyd from the
+// environment.
+func New() (*SkyDB, error) {
+	sk, err1 := base64.StdEncoding.DecodeString(os.Getenv("SKYDB_SEC_KEY"))
+	pk, err2 := base64.StdEncoding.DecodeString(os.Getenv("SKYDB_PUB_KEY"))
+	if err1 != nil || err2 != nil {
+		return nil, errors.AddContext(errors.Compose(err1, err2), "failed to decode SKYDB_SEC_KEY and/or SKYDB_PUB_KEY")
+	}
+	if len(sk) == 0 || len(pk) == 0 {
+		return nil, errors.New("missing SKYDB_SEC_KEY or SKYDB_PUB_KEY environment variable")
+	}
+	opts, err := client.DefaultOptions()
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to get default client options")
+	}
+	skydEndpoint := os.Getenv("SKYDB_ENDPOINT")
+	if skydEndpoint == "" {
+		return nil, errors.New("missing SKYDB_ENDPOINT environment variable")
+	}
+	opts.Address = skydEndpoint
+	skydb := &SkyDB{Client: &client.Client{opts}}
+	copy(skydb.sk[:], sk)
+	copy(skydb.pk[:], pk)
+	return skydb, nil
+}
+
+// Read retrieves from SkyDB the data that corresponds to the given key set.
+func (db SkyDB) Read(dataKey crypto.Hash) ([]byte, uint64, error) {
+	s, rev, err := registryRead(db.Client, db.pk, dataKey)
```

Comment:
f/u: Can use the SkyDB functions from the Go SDK, once that's ready -- one day. 😛

Reply:
Once that's ready - definitely. :D I wanted to use it already but... you know.. it's not there.

OTOH, we can use these functions here as the initial versions of the ones in the SDK.